File: ios/SampleUI/Sources/SampleUI/Components/UseSmileIDSampleNavBar.swift
import SwiftUI

/// The three destinations the nav bar switches between. The token affordance is not one of them.
public enum UseSmileIDSampleNavItem: String, CaseIterable, Sendable {
  case products, verifications, settings

  public var label: String {
    rawValue.prefix(1).uppercased() + rawValue.dropFirst()
  }

  public var testId: String {
    switch self {
    case .products: UseSmileIDSampleTestIds.navProducts
    case .verifications: UseSmileIDSampleTestIds.navVerifications
    case .settings: UseSmileIDSampleTestIds.navSettings
    }
  }

  public var icon: SmileIcon {
    switch self {
    case .products: SmileIcons.products
    case .verifications: SmileIcons.verifications
    case .settings: SmileIcons.settings
    }
  }
}

/// The countdown ring: green rather than primary, driven by remaining time rather than a duration.
public struct UseSmileIDSampleTokenRing: View {
  private let progress: Double

  public init(progress: Double) {
    self.progress = progress
  }

  public var body: some View {
    ZStack {
      Circle().stroke(smileTokenRing.opacity(smileTokenRingTrackOpacity), lineWidth: Self.width)
      Circle()
        .trim(from: 0, to: min(max(progress, 0), 1))
        .stroke(smileTokenRing, style: StrokeStyle(lineWidth: Self.width, lineCap: .round))
        .rotationEffect(.degrees(-90))
    }
  }

  private static var width: CGFloat {
    3
  }
}

/// Three tabs plus a detached token button that navigates rather than switching tab; the pill has its own fill, per the `navBarFill` delta.
public struct UseSmileIDSampleNavBar: View {
  private let selected: UseSmileIDSampleNavItem
  private let sessionProgress: Double?
  private let onSelect: (UseSmileIDSampleNavItem) -> Void
  private let onToken: () -> Void

  @ScaledMetric(relativeTo: .body) private var tokenSize: CGFloat = 56
  @Environment(\.useSmileIDSampleColors) private var colors

  public init(
    selected: UseSmileIDSampleNavItem,
    sessionProgress: Double? = nil,
    onSelect: @escaping (UseSmileIDSampleNavItem) -> Void,
    onToken: @escaping () -> Void
  ) {
    self.selected = selected
    self.sessionProgress = sessionProgress
    self.onSelect = onSelect
    self.onToken = onToken
  }

  public var body: some View {
    HStack(spacing: SmileSpacing.spacingXs) {
      HStack(spacing: 0) {
        ForEach(UseSmileIDSampleNavItem.allCases, id: \.self) { item in
          tab(item)
        }
      }
      .padding(.horizontal, SmileSpacing.spacingXs)
      .padding(.vertical, SmileSpacing.spacingXxs)
      .frame(maxWidth: .infinity)
      .background(Capsule().fill(colors.navBar).shadow(radius: Self.elevation))
      token
    }
    .padding(.horizontal, SmileSpacing.spacingMd)
    .padding(.vertical, SmileSpacing.spacingSm)
  }

  private func tab(_ item: UseSmileIDSampleNavItem) -> some View {
    // Icon above label, primary when active — `spec/components.json`, corrected 2026-08-17.
    let tint = item == selected ? colors.primary : colors.foreground
    return Button { onSelect(item) } label: {
      VStack(spacing: SmileSpacing.spacingXxs) {
        UseSmileIDSampleIcon(item.icon, tint: tint, size: Self.tabIconSize)
        UseSmileIDSampleText(item.label, style: UseSmileIDSampleTheme.type.tabFont)
          .foregroundColor(tint)
          .multilineTextAlignment(.center)
      }
      // Padded, not floored at `sizeControlMd`: 44 is shorter than the content, so it shrank the pill.
      .frame(maxWidth: .infinity)
      .padding(.vertical, SmileSpacing.spacingXs)
      .contentShape(Rectangle())
    }
    .buttonStyle(.plain)
    .accessibilityAddTraits(item == selected ? [.isButton, .isSelected] : .isButton)
    .useSmileIDSampleTestId(item.testId)
  }

  /// The ring is painted outside the button's bounds: sized into the layout it pushed the bar off a 393pt screen.
  private var token: some View {
    Button(action: onToken) {
      VStack(spacing: 0) {
        UseSmileIDSampleIcon(SmileIcons.tokenScan, tint: colors.foreground, size: SmileSpacing.sizeIconSm)
        UseSmileIDSampleText("Token", style: UseSmileIDSampleTheme.type.textStyleOverline.with(size: 9))
          .foregroundColor(colors.foreground)
      }
      .frame(minWidth: tokenSize, minHeight: tokenSize)
      .background(Circle().fill(colors.navBar).shadow(radius: Self.elevation))
      .overlay {
        if let sessionProgress {
          UseSmileIDSampleTokenRing(progress: sessionProgress).padding(-Self.ringBleed)
        }
      }
    }
    .buttonStyle(.plain)
    .useSmileIDSampleTestId(UseSmileIDSampleTestIds.navToken)
  }

  private static var elevation: CGFloat {
    8
  }

  private static var ringBleed: CGFloat {
    3
  }

  /// The Compose twin's own 21, which no scale token carries.
  private static var tabIconSize: CGFloat {
    21
  }
}

File: flutter/sample_ui/lib/src/components/use_smileid_sample_nav_bar.dart
import 'dart:math' as math;

import 'package:flutter/material.dart';

import '../theme/use_smileid_sample_colors.dart';
import '../theme/use_smileid_sample_theme.dart';
import '../theme/use_smileid_sample_typography.dart';
import '../tokens/smile_icons.dart';
import '../tokens/smile_product_hues.dart';
import '../tokens/smile_tokens.dart';
import '../use_smileid_sample_test_ids.dart';
import 'use_smileid_sample_icon.dart';

/// The three destinations the nav bar switches between. The token affordance is not one of them.
enum UseSmileIDSampleNavItem {
  /// The products grid.
  products(
    UseSmileIDSampleTestIds.navProducts,
    'Products',
    SmileIcons.products,
  ),

  /// The verifications list.
  verifications(
    UseSmileIDSampleTestIds.navVerifications,
    'Verifications',
    SmileIcons.verifications,
  ),

  /// Settings.
  settings(
    UseSmileIDSampleTestIds.navSettings,
    'Settings',
    SmileIcons.settings,
  );

  const UseSmileIDSampleNavItem(this.testId, this.label, this.icon);

  /// The `sample_*` id a flow taps.
  final String testId;

  /// The tab's label, beneath its icon.
  final String label;

  /// The tab's mark, which the design supplies.
  final String icon;
}

/// A floating pill of three tabs, plus a detached token button that navigates rather than
/// switching tab.
///
/// The bar FLOATS over the content: the list scrolls underneath it and the background stays
/// continuous, so a screen's own trailing spacer is what lets its last row scroll clear.
class UseSmileIDSampleNavBar extends StatelessWidget {
  /// [sessionProgress] drives the ring, 1 fresh to 0 expired, from the deadline not an animation.
  const UseSmileIDSampleNavBar({
    required this.selected,
    required this.onSelect,
    required this.onTokenTap,
    this.sessionProgress,
    super.key,
  });

  /// Which tab is active.
  final UseSmileIDSampleNavItem selected;

  /// Switches tab.
  final ValueChanged<UseSmileIDSampleNavItem> onSelect;

  /// Opens the token screen, which is a push rather than a tab.
  final VoidCallback onTokenTap;

  /// The session's remaining fraction, absent when there is no session.
  final double? sessionProgress;

  @override
  Widget build(BuildContext context) {
    final UseSmileIDSampleColors colors = UseSmileIDSampleTheme.colorsOf(
      context,
    );
    return Padding(
      // Edge to edge, so without its own inset the bar sits under the system navigation bar.
      padding: EdgeInsets.only(
        left: SmileDimens.spacingMd,
        right: SmileDimens.spacingMd,
        top: SmileDimens.spacingSm,
        bottom:
            SmileDimens.spacingSm + MediaQuery.viewPaddingOf(context).bottom,
      ),
      child: Row(
        children: <Widget>[
          Expanded(
            child: Material(
              // The pill has its own fill: the design recesses it below the page, which this
              // app's page colour cannot express without hiding the bar.
              color: colors.navBar,
              shape: const StadiumBorder(),
              elevation: _barElevation,
              child: Padding(
                padding: const EdgeInsets.symmetric(
                  horizontal: SmileDimens.spacingXs,
                  vertical: SmileDimens.spacingXxs,
                ),
                child: Row(
                  children: <Widget>[
                    for (final UseSmileIDSampleNavItem item
                        in UseSmileIDSampleNavItem.values)
                      Expanded(
                        child: _NavBarTab(
                          item: item,
                          selected: item == selected,
                          onTap: () => onSelect(item),
                        ),
                      ),
                  ],
                ),
              ),
            ),
          ),
          const SizedBox(width: SmileDimens.spacingXs),
          _TokenAffordance(progress: sessionProgress, onTap: onTokenTap),
        ],
      ),
    );
  }
}

/// The countdown ring: green rather than primary, and driven by remaining time.
class UseSmileIDSampleTokenRing extends StatelessWidget {
  /// [progress] is 1 fresh to 0 expired; anything outside that is clamped rather than rejected.
  const UseSmileIDSampleTokenRing({
    required this.progress,
    this.size,
    super.key,
  });

  /// The session's remaining fraction.
  final double progress;

  /// The square the ring is drawn into.
  final double? size;

  @override
  Widget build(BuildContext context) => CustomPaint(
    size: Size.square(size ?? _tokenSize),
    painter: _TokenRingPainter(progress: progress, inflate: 0),
  );
}

/// The ring is painted OUTSIDE the button's bounds rather than laid out around it: given its own
/// room in the layout it pushed the token button off a 393-wide screen.
class _TokenAffordance extends StatelessWidget {
  const _TokenAffordance({required this.progress, required this.onTap});

  final double? progress;
  final VoidCallback onTap;

  @override
  Widget build(BuildContext context) {
    final UseSmileIDSampleColors colors = UseSmileIDSampleTheme.colorsOf(
      context,
    );
    final Widget button = Material(
      color: colors.navBar,
      shape: const CircleBorder(),
      elevation: _barElevation,
      child: InkWell(
        onTap: onTap,
        customBorder: const CircleBorder(),
        child: ConstrainedBox(
          // A minimum, not a fixed box: a fixed 58 ellipsises "Token" at 2x.
          constraints: const BoxConstraints(
            minWidth: _tokenSize,
            minHeight: _tokenSize,
          ),
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            mainAxisSize: MainAxisSize.min,
            children: <Widget>[
              UseSmileIDSampleIcon(
                asset: SmileIcons.tokenScan,
                tint: colors.foreground,
                size: SmileDimens.sizeIconSm,
              ),
              const SizedBox(height: SmileDimens.spacingXxs),
              Text(
                'Token',
                style: UseSmileIDSampleType.textStyleOverline.copyWith(
                  fontSize: _tokenLabelSize,
                  color: colors.foreground,
                ),
              ),
            ],
          ),
        ),
      ),
    );
    return Semantics(
      identifier: UseSmileIDSampleTestIds.navToken,
      button: true,
      child: progress == null
          ? button
          : CustomPaint(
              painter: _TokenRingPainter(
                progress: progress!,
                inflate: _ringBleed,
              ),
              child: button,
            ),
    );
  }
}

class _TokenRingPainter extends CustomPainter {
  const _TokenRingPainter({required this.progress, required this.inflate});

  final double progress;
  final double inflate;

  @override
  void paint(Canvas canvas, Size size) {
    final Paint stroke = Paint()
      ..style = PaintingStyle.stroke
      ..strokeWidth = _ringThickness
      ..strokeCap = StrokeCap.round;
    final double offset = _ringThickness / 2 - inflate;
    final double diameter = size.shortestSide - _ringThickness + inflate * 2;
    final Rect bounds = Rect.fromLTWH(offset, offset, diameter, diameter);

    canvas.drawArc(
      bounds,
      0,
      _fullTurn,
      false,
      stroke
        ..color = smileTokenRing.withValues(alpha: smileTokenRingTrackOpacity),
    );
    canvas.drawArc(
      bounds,
      _quarterTurnUp,
      _fullTurn * progress.clamp(0, 1),
      false,
      stroke..color = smileTokenRing,
    );
  }

  @override
  bool shouldRepaint(_TokenRingPainter oldDelegate) =>
      oldDelegate.progress != progress || oldDelegate.inflate != inflate;
}

class _NavBarTab extends StatelessWidget {
  const _NavBarTab({
    required this.item,
    required this.selected,
    required this.onTap,
  });

  final UseSmileIDSampleNavItem item;
  final bool selected;
  final VoidCallback onTap;

  @override
  Widget build(BuildContext context) {
    final UseSmileIDSampleColors colors = UseSmileIDSampleTheme.colorsOf(
      context,
    );
    // Unselected takes the warm Off_black, not muted text, which is what the frame draws.
    final Color tint = selected ? colors.primary : colors.foreground;
    return Semantics(
      identifier: item.testId,
      selected: selected,
      button: true,
      child: InkWell(
        onTap: onTap,
        customBorder: const StadiumBorder(),
        child: Padding(
          padding: const EdgeInsets.all(SmileDimens.spacingXs),
          child: Column(
            mainAxisSize: MainAxisSize.min,
            children: <Widget>[
              UseSmileIDSampleIcon(
                asset: item.icon,
                tint: tint,
                size: _tabIconSize,
              ),
              const SizedBox(height: SmileDimens.spacingXxs),
              Text(
                item.label,
                textAlign: TextAlign.center,
                style: UseSmileIDSampleType.textStyleOverline.copyWith(
                  color: tint,
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

/// A spacing token standing in for an elevation, because no token carries one — see the
/// `floatingElevation` delta. It matches the design's 8 y-offset, which is the reproducible half.
const double _barElevation = SmileDimens.space8;

/// How far the ring is pushed outside the button it circles.
const double _ringBleed = 5;

/// The ring's own stroke.
const double _ringThickness = 4;

/// The tab mark, read off the board rather than the icon scale.
const double _tabIconSize = 21;

/// The token button's minimum, which is NOT size.control-md's 44.
const double _tokenSize = 58;

/// The token label's run.
const double _tokenLabelSize = 8.5;

const double _fullTurn = 360 * math.pi / 180;
const double _quarterTurnUp = -90 * math.pi / 180;
